## 4.7 Deriving New Variables

### Steps for this task:

#### 1. Import ords_prods_merge dataframe

#### 2. Create subset working with 1,000,000 rows

#### 3. Define function that will be used to create new column called price_label

#### 4. If-Statements with the loc() Function for price range column

#### 5. If-Statements with For-Loops to create new column based on busiest day

#### 6. Create New for-loop based on busiest and slowest days

#### 7. Create new column "busiest_period_of_day" to identify time periods “Most orders,” “Average orders,” and “Fewest orders.”

#### 8. Export ords_prods_merge dataframe into pickle file

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [10]:
# Create file path
path = r'C:\Users\gre60\Desktop\30-05-2024 Instacart Basket Analysis'

In [2]:
path_2 = r'/Users/gre604/Desktop/30-05-2024 Instacart Basket Analysis'

### 1. Import ords_prods_merge dataframe

In [3]:
# Importing ords_prods_merge pickle file into df
ords_prods_merge = pd.read_pickle(os.path.join(path_2, 'Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [4]:
# Check output
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order first_order  add_to_cart_order  reordered _merge  
0                     3.0          no                  5          0   both  
1                    20.0          no                  1          1   both  
2                     6.0          no                 20          0   both  
3                     NaN         yes                 10          0   both  
4                     9.0          no                 11          1   both

In [5]:
# Check dimensions
ords_prods_merge.shape

(32404859, 15)

### 2. Create subset working with 1,000,000 rows

In [6]:
# Creating subset
df = ords_prods_merge[:1000000]

In [7]:
# Checking subset df results
df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order first_order  add_to_cart_order  reordered _merge  
0                     3.0          no                  5          0   both  
1                    20.0          no                  1          1   both  
2                     6.0          no                 20          0   both  
3                     NaN         yes                 10          0   both  
4                     9.0          no                 11          1   both

In [8]:
# Checking subset dimensions
df.shape

(1000000, 15)

### 3. Define function that will be used to create new column called price_label

In [9]:
# Creating function to create price label
def price_label(row):
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High-range'
    else: return 'Not enough data'

In [10]:
# Apply function
df['price_range'] = df.apply(price_label, axis = 1)

<ipython-input-10-c285306f3b3b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [11]:
df['price_range'].value_counts(dropna = False)

Mid-range product    652638
Low-range product    338018
High-range             9344
Name: price_range, dtype: int64

In [12]:
df['prices'].max()

24.5

### 4. If-Statements with the loc() Function

In [13]:
# Use loc() method on df subset to apply price range function
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'

/Users/gre604/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1596: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = _infer_fill_value(value)
/Users/gre604/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


In [14]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5) , 'price_range_loc'] = 'Mid range product'

/Users/gre604/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


In [15]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low range product'

/Users/gre604/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


In [16]:
df['price_range_loc'].value_counts(dropna = False)

Mid range product     652638
Low range product     338018
High range product      9344
Name: price_range_loc, dtype: int64

In [17]:
# Now use loc() for entire dataframe
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High range product'

In [18]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5) , 'price_range_loc'] = 'Mid range product'

In [19]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low range product'

In [20]:
# Check results with entire dataframe
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

Mid range product     21860860
Low range product     10126321
High range product      417678
Name: price_range_loc, dtype: int64

### 5. If-Statements with For-Loops to create new column based on busiest days

In [21]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [22]:
# Creating for-loop for new busiet day column
result = []

for value in ords_prods_merge['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [23]:
# Checking results of for-loop
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [24]:
# Adding busiest day column to dataframe
ords_prods_merge['busiest_day'] = result

In [25]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

### 6. Create New for-loop based on busiest and slowest days

In [26]:
#Creating new for-loop based on busiest and slowest days

new_result = []

for value in ords_prods_merge['orders_day_of_week']:
    if value == 0 or value == 1:
        new_result.append('Busiest days')
    elif value == 4 or value == 3:
        new_result.append('Slowest days')
    else:
        new_result.append('Regularly busy')

In [27]:
new_result

['Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',

In [28]:
# Adding busiest days for-loop to datafram
ords_prods_merge['busiest_days'] = new_result

In [29]:
ords_prods_merge['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

##### The total value counts for the ords_prods_merge dataframe is equal to the total sum of all groupings (i.e. "Regularly busy" + "Busiest days" + "Slowest days"). In addition, the listed sums for each label match the sum totals for the qualifying days of the week (e.g. Total "Busiest days" = 11864412, which matches orders_day_of_week 0 + 1, or 6204182 + 5660230).

### 7. Create new column "busiest_period_of_day" to identify time periods “Most orders,” “Average orders,” and “Fewest orders.”

In [30]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order first_order  add_to_cart_order  reordered _merge  \
0                     3.0          no                  5          0   both   
1                    20.0          no                  1          1   both   
2                     6.0          no                 20          0   both   
3                     NaN         yes                 10          0   both   
4                     9.0          no                 11          1   both   

     price_range_loc     busiest_day    busiest_days  
0  Mid range product  Regularly busy  Regularly busy  
1  Mid range product  Regularly busy  Regularly busy  
2  Mid range product     Busiest day    Busiest days  
3  Mid range product  Regularly busy    Slowest days  
4  Mid range product      Least busy    Slowest days

In [31]:
ords_prods_merge['order_hour_of_day'].value_counts()

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [32]:
# Create for-loop to identify time periods that are busiest

result_2 = []

for value in ords_prods_merge['order_hour_of_day']:
    if value in [10,11,14,15,13,12,16,9]:
        result_2.append('Most orders')
    elif value in [17,8,18,19,20,7,21,22]:
        result_2.append('Average orders')
    else:
        result_2.append('Fewest orders')

In [33]:
# Adding busiest period of day column
ords_prods_merge['busiest_period_of_day'] = result_2

In [34]:
# Print freqeuncy of busies_period_of_day column
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

### 8. Export ords_prods_merge dataframe into pickle file

In [90]:
ords_prods_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

In [35]:
ords_prods_merge.to_pickle(os.path.join(path_2, 'Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))